File: src/java/com/lightningtoads/toadlet/hop/Constraint.java
/********** Copyright header - do not remove **********
 *
 * The Toadlet Engine
 *
 * Copyright 2009, Lightning Toads Productions, LLC
 *
 * Author(s): Alan Fischer, Andrew Fischer
 *
 * This file is part of The Toadlet Engine.
 *
 * The Toadlet Engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * The Toadlet Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public License
 * along with The Toadlet Engine.  If not, see <http://www.gnu.org/licenses/>.
 *
 ********** Copyright header - do not remove **********/

package com.lightningtoads.toadlet.hop;

#include <com/lightningtoads/toadlet/hop/Types.h>

public final class Constraint{
	public Constraint(){}

	public Constraint(Solid startSolid,Solid endSolid){
		reset();

		setStartSolid(startSolid);
		setEndSolid(endSolid);
	}
	
	public Constraint(Solid startSolid,Vector3 endPoint){
		reset();

		setStartSolid(startSolid);
		setEndPoint(endPoint);
	}

	public void reset(){
		destroy();
		
		mSpringConstant=Math.ONE;
		mDampingConstant=Math.ONE;
		mDistanceThreshold=Math.ONE;

		mSimulator=null;
	}

	public void destroy(){
		if(mStartSolid!=null){
			mStartSolid.activate();
			mStartSolid.internal_removeConstraint(this);
			mStartSolid=null;
		}

		if(mEndSolid!=null){
			mEndSolid.activate();
			mEndSolid.internal_removeConstraint(this);
			mEndSolid=null;
		}
	}

	public void setStartSolid(Solid startSolid){
		if(mEndSolid!=null){
			mEndSolid.activate();
		}

		if(mStartSolid!=null){
			mStartSolid.activate();
			mStartSolid.internal_removeConstraint(this);
			mStartSolid=null;
		}

		if(startSolid!=null){
			startSolid.internal_addConstraint(this);
			startSolid.activate();
			mStartSolid=startSolid;
		}
	}

	public Solid getStartSolid(){return mStartSolid;}

	public void setEndSolid(Solid endSolid){
		if(mStartSolid!=null){
			mStartSolid.activate();
		}

		if(mEndSolid!=null){
			mEndSolid.activate();
			mEndSolid.internal_removeConstraint(this);
			mEndSolid=null;
		}

		if(endSolid!=null){
			endSolid.internal_addConstraint(this);
			endSolid.activate();
			mEndSolid=endSolid;
		}
	}

	public Solid getEndSolid(){return mEndSolid;}

	public void setEndPoint(Vector3 endPoint){
		if(mStartSolid!=null){
			mStartSolid.activate();
		}

		if(mEndSolid!=null){
			mEndSolid.activate();
			mEndSolid.internal_removeConstraint(this);
			mEndSolid=null;
		}

		mEndPoint.set(endPoint);
	}

	Vector3 getEndPoint(){return mEndPoint;}

	public void setSpringConstant(scalar springConstant){mSpringConstant=springConstant;}
	public scalar getSpringConstant(){return mSpringConstant;}

	public void setDampingConstant(scalar dampingConstant){mDampingConstant=dampingConstant;}
	public scalar getDampingConstant(){return mDampingConstant;}

	public void setDistanceThreshold(scalar threshold){mDistanceThreshold=threshold;}
	public scalar getDistanceThreshold(){return mDistanceThreshold;}

	public void activate(){}
	public boolean getActive(){return (mSimulator!=null);}

	public Simulator getSimulator(){return mSimulator;}

	void internal_setSimulator(Simulator s){mSimulator=s;}

	Solid mStartSolid=null;
	Solid mEndSolid=null;
	Vector3 mEndPoint=new Vector3();

	scalar mSpringConstant=0;
	scalar mDampingConstant=0;
	scalar mDistanceThreshold=0;
	
	Simulator mSimulator=null;
}

File: src/cpp/toadlet/hop/Constraint.cpp
/********** Copyright header - do not remove **********
 *
 * The Toadlet Engine
 *
 * Copyright 2009, Lightning Toads Productions, LLC
 *
 * Author(s): Alan Fischer, Andrew Fischer
 *
 * This file is part of The Toadlet Engine.
 *
 * The Toadlet Engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * The Toadlet Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public License
 * along with The Toadlet Engine.  If not, see <http://www.gnu.org/licenses/>.
 *
 ********** Copyright header - do not remove **********/

#include <toadlet/hop/Constraint.h>
#include <toadlet/hop/Solid.h>

namespace toadlet{
namespace hop{

Constraint::Constraint():
	//mStartSolid,
	//mEndSolid,
	//mEndPoint,

	mSpringConstant(0),
	mDampingConstant(0),
	mDistanceThreshold(0),

	mSimulator(NULL)
{
	reset();
}

Constraint::Constraint(Solid::ptr startSolid,Solid::ptr endSolid):
	//mStartSolid,
	//mEndSolid,
	//mEndPoint,

	mSpringConstant(0),
	mDampingConstant(0),
	mDistanceThreshold(0),

	mSimulator(NULL)
{
	reset();

	setStartSolid(startSolid);
	setEndSolid(endSolid);
}

Constraint::Constraint(Solid::ptr startSolid,const Vector3 &endPoint):
	//mStartSolid,
	//mEndSolid,
	//mEndPoint,

	mSpringConstant(0),
	mDampingConstant(0),
	mDistanceThreshold(0),

	mSimulator(NULL)
{
	reset();

	setStartSolid(startSolid);
	setEndPoint(endPoint);
}

Constraint::~Constraint(){
	destroy();
}

void Constraint::reset(){
	destroy();

	mSpringConstant=Math::ONE;
	mDampingConstant=Math::ONE;
	mDistanceThreshold=Math::ONE;

	mSimulator=NULL;
}

void Constraint::destroy(){
	if(mStartSolid!=NULL){
		mStartSolid->activate();
		mStartSolid->internal_removeConstraint(this);
		mStartSolid=NULL;
	}

	if(mEndSolid!=NULL){
		mEndSolid->activate();
		mEndSolid->internal_removeConstraint(this);
		mEndSolid=NULL;
	}
}

void Constraint::setStartSolid(Solid::ptr startSolid){
	if(mEndSolid!=NULL){
		mEndSolid->activate();
	}

	if(mStartSolid!=NULL){
		mStartSolid->activate();
		mStartSolid->internal_removeConstraint(this);
		mStartSolid=NULL;
	}

	if(startSolid!=NULL){
		startSolid->internal_addConstraint(this);
		startSolid->activate();
		mStartSolid=startSolid;
	}
}

void Constraint::setEndSolid(Solid::ptr endSolid){
	if(mStartSolid!=NULL){
		mStartSolid->activate();
	}

	if(mEndSolid!=NULL){
		mEndSolid->activate();
		mEndSolid->internal_removeConstraint(this);
		mEndSolid=NULL;
	}

	if(endSolid!=NULL){
		endSolid->internal_addConstraint(this);
		endSolid->activate();
		mEndSolid=endSolid;
	}
}

void Constraint::setEndPoint(const Vector3 &endPoint){
	if(mStartSolid!=NULL){
		mStartSolid->activate();
	}

	if(mEndSolid!=NULL){
		mEndSolid->activate();
		mEndSolid->internal_removeConstraint(this);
		mEndSolid=NULL;
	}

	mEndPoint.set(endPoint);
}

void Constraint::setSpringConstant(scalar springConstant){
	mSpringConstant=springConstant;
}

void Constraint::setDampingConstant(scalar dampingConstant){
	mDampingConstant=dampingConstant;
}

void Constraint::setDistanceThreshold(scalar threshold){
	mDistanceThreshold=threshold;
}

void Constraint::internal_setSimulator(Simulator *s){
	mSimulator=s;
}

}
}

